### PR TITLE
User can read a specific article 

### PR DESCRIPTION
**Pull request by DIzosimov**
[PT Link](https://www.pivotaltracker.com/story/show/169253715)

## PR Content
- Adds ViewArticle component to view single article
- Adds happy and sad path testing for viewing single article

## Screenshots
Screenshot 2019-11-01 at 16 13 36

### CODE REVIEW

**Review by miyeaier — APPROVED** [review]

File: cypress/fixtures/articles.json [hunk 1]
```diff
@@ -14,5 +14,13 @@
     "author": "Some awesome author",
     "image": "https://t3.ftcdn.net/jpg/02/48/42/64/240_F_248426448_NVKLywWqArG2ADUxDq6QprtIzsF82dMF.jpg",
     "publish_date": "20 October 2019"
+  },
+  {
+    "id": 3,
+    "title": "Fuc u ",
```
Comment (piavW) [inline cypress/fixtures/articles.json]: rename this

File: src/Components/ListArticles.jsx [hunk 2]
```diff
@@ -1,89 +1,105 @@
...
+    let renderListArticles;
+    let renderArticle = this.state.renderArticle
+    const articleData = this.state.articles
+    let specificArticle
+    let error_message
+
+    if (this.state.error_message) {
+      error_message = <p>{ this.state.error_message }</p>
+    }
+  if (articleData.length !== 0) {
```
Comment (piavW) [inline src/Components/ListArticles.jsx]: indentation?

File: src/Components/ViewArticle.jsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  async componentDidMount() {
+    let response = await getArticle(this.props.chosenArticle)
+    if (response.status === 200) {
+      this.setState({
+        article: response.data
+      })
+    } else {
+      this.props.renderErrorMessage(response) // Check this renderErrorMessage call
```
Comment (piavW) [inline src/Components/ViewArticle.jsx]: remove comments

File: src/Components/ViewArticle.jsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    let singleArticle
+    const article = this.state.article
+
+    if (article !== null) {
+      singleArticle = (
+        <div id="single-article">
+          <Container>
+            <Item.Group> 
+              <Item>
+                {/* <Item.Image size='tiny' src='https://react.semantic-ui.com/images/wireframe/image.png' /> */}
```
Comment (piavW) [inline src/Components/ViewArticle.jsx]: commented out code

**Review by piavW — CHANGES_REQUESTED** [review]
Remove commented lines and see my comments

**Review by yasmineezequiel — APPROVED** [review]

**Review by piavW — APPROVED** [review]

**Review by yasmineezequiel — APPROVED** [review]

**Review by miyeaier — APPROVED** [review]

**Review by beccaburns — APPROVED** [review]

File: src/Components/ViewArticle.jsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    } else {
+      this.props.renderErrorMessage(response)
+    }
+  }
+
+  render() {
+    let singleArticle
+    const article = this.state.article
+
+    if (article !== null) {
```
Comment (oliverochman) [inline src/Components/ViewArticle.jsx]: please refactor
`if (article) {`

File: cypress/integration/userCanViewArticle.spec.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  it('sees error message for chosen non-existent article', () => {
+    cy.server()
+    cy.route({
+      method: 'GET',
+      url: 'http://localhost:3000/api/v1/articles/1',
+      status: 404,
+      response: 'fixture:unsuccessfully_viewing_article.json'
+    })
+    cy.visit('http://localhost:3001')
+    cy.contains('Network Error')
```
Comment (oliverochman) [inline cypress/integration/userCanViewArticle.spec.js]: what specific element contains this?
shouldn't you see the error message that you have in `fixture:unsuccessfully_viewing_article.json`?

File: cypress/integration/userCanSignUp.spec.js [hunk 7]
```diff
@@ -51,6 +51,6 @@ describe('User can sign up', () => {
       cy.get('#password-confirmation').type('wrong_password')
     })
     cy.get('#submit-signup-form').click()
-    cy.get('#error-message').should('contain', 'Invalid credentials')
+    cy.contains('Invalid credentials')
```
Comment (oliverochman) [inline cypress/integration/userCanSignUp.spec.js:54]: why did you remove `cy.get('#error-message')`?

File: package.json [hunk 8]
```diff
@@ -7,6 +7,7 @@
     "react": "^16.11.0",
     "react-dom": "^16.11.0",
     "react-redux": "^7.1.1",
+    "react-router-dom": "^5.1.2",
```
Comment (oliverochman) [inline package.json:12]: do you need this?

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: cypress/integration/userCanSignUp.spec.js (hunk 7 above)
Comment (DIzosimov) [reply cypress/integration/userCanSignUp.spec.js:54]: I've inspected the element and tried bunch of different ways but for some reason Invalid Credentials cant be found inside of the #error-message even though it is displayed.

File: package.json (hunk 8 above)
Comment (DIzosimov) [reply package.json:12]: We were going to implement react-router-dom but are moving it to the navbar feature instead. Should we remove it from the package.json?

**Review by piavW — APPROVED** [review]

File: package.json (hunk 8 above)
Comment (oliverochman) [reply package.json:12]: yes

File: cypress/integration/userCanViewArticle.spec.js (hunk 6 above)
Comment (oliverochman) [reply cypress/integration/userCanViewArticle.spec.js]: ???

Comment (DIzosimov) [reply cypress/integration/userCanViewArticle.spec.js]: I can't make it display the error message nor the right one nor the fixture file.. :)

File: package.json (hunk 8 above)
Comment (faraznaeem) [reply package.json:12]: Please address the changes

**Review by faraznaeem — CHANGES_REQUESTED** [review]

File: package.json (hunk 8 above)
Comment (yasmineezequiel) [reply package.json:12]: Fixed

**Review by yasmineezequiel — APPROVED** [review]

File: cypress/integration/userCanViewListedArticles.spec.js [hunk 9]
```diff
@@ -23,9 +21,10 @@ describe('user can view listed articles', () => {
     cy.route({
       method: 'GET',
       url: 'http://localhost:3000/api/v1/articles',
-      status: 400,
+      status: 404,
+      response: 'fixture:unsuccessfully_viewing_articles.json'
     })
     cy.visit('http://localhost:3001')
-    cy.contains('Network Error')
+    cy.contains('Request failed with status code 404')
```
Comment (GergKllai1) [inline cypress/integration/userCanViewListedArticles.spec.js:28]: That is the error message you show to the user?

File: src/Components/ListArticles.jsx [hunk 10]
```diff
@@ -1,89 +1,106 @@
...
 
-    if (this.state.error_message) {
-      return(
-        <div>
-          { this.state.error_message }
-        </div>
-      )
-    }
+  render() {
+    let renderListArticles;
```
Comment (GergKllai1) [inline src/Components/ListArticles.jsx]: You can declare multiple variables in one line

File: src/Components/Signup.jsx [hunk 11]
```diff
@@ -73,7 +73,7 @@ class Signup extends Component {
       }
     }
     if (this.state.errorMessage !== '') {
```
Comment (GergKllai1) [inline src/Components/Signup.jsx]: Javascript returns false if you have an empty string.
You can shorten this by saying `if(this.state.errorMessage)`

File: src/Components/ListArticles.jsx [hunk 12]
```diff
@@ -1,89 +1,106 @@
...
+                      </Item.Content>
+                    </Item>
+                  </Item.Group> 
+                </div>
+              })}
+            </>
+          )
+        }
+      }
+      if (renderArticle === true) {
```
Comment (GergKllai1) [inline src/Components/ListArticles.jsx]: You don't have to add `=== true`

File: src/Components/ListArticles.jsx [hunk 13]
```diff
@@ -1,89 +1,106 @@
...
-          renderListArticles = (
-          <div>
-            {this.state.error_message}
-          </div>
+    if (this.state.error_message) {
+      error_message = <p>{ this.state.error_message }</p>
+    }
+
+    if (articleData.length !== 0) {
+        if (renderArticle === false) {
```
Comment (GergKllai1) [inline src/Components/ListArticles.jsx]: You can shorten this: `if(!renderArticle)`

File: src/Components/ListArticles.jsx [hunk 14]
```diff
@@ -1,89 +1,106 @@
...
-        error_message: result.error_message
-      })
-    } else {
-      this.setState({
-        articles: result
-      })
-    }
-  }
+  async getArticles() {
+    let result = await getData()
```
Comment (GergKllai1) [inline src/Components/ListArticles.jsx:31]: Why are you extracting out the `getData()` function? You double the error handling because you do that in the module and you do that in here as well.
I see that you do that multiple times in the code, so we can go with this for now, but I would rethink this approach because you are adding additional bloat to the codebase.

**Review by GergKllai1 — CHANGES_REQUESTED** [review]

**Review by piavW — APPROVED** [review]

**Review by yasmineezequiel — APPROVED** [review]

**Review by miyeaier — APPROVED** [review]

File: src/Components/ViewArticle.jsx [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+import { Container, Item } from 'semantic-ui-react'
+import { getArticle } from '../Modules/RequestArticles'
+
+class ViewArticle extends Component {
+  state = {
+    article: null
+  }
+
+  async componentDidMount() {
+  debugger
```
Comment (oliverochman) [inline src/Components/ViewArticle.jsx]: please remove the debugger

**Review by oliverochman — CHANGES_REQUESTED** [review]
